[buildbot] Add de-duper script for results

JonWBedard opened the pull request:
#### f07efecdac0917ed92543b28e9af8c2e0305ba10
<pre>
[buildbot] Add de-duper script for results
<a href="https://bugs.webkit.org/show_bug.cgi?id=281334">https://bugs.webkit.org/show_bug.cgi?id=281334</a>
<a href="https://rdar.apple.com/137770187">rdar://137770187</a>

Reviewed by Sam Sneddon.

Attempt to hard-link identical PNG files to a shared &quot;storage&quot; directory.

* Tools/CISupport/Shared/de-duper: Added.
(parse_args): Caller can specify root directory and storage directory.
(sha1): Compute the sha1 hash of the specified file.
(main): Iterate through all PNG files in the root directory, if those files
are already hard-links, skip them. Otherwise, compute the sha1 hash of the source
PNG files and move the file content to the storage directory and hard-link to that.

Canonical link: <a href="https://commits.webkit.org/285362@main">https://commits.webkit.org/285362@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a231377e06fc207b9cb88cf0356598b43dedb9b2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/72419 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/51840 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/25207 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/76600 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/23628 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/74534 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/59644 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/23450 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/76600 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/23628 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/75486 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/46937 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/62366 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/76600 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/43588 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/19833 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/21978 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/65469 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/20191 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/78267 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/87/builds/16660 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/19327 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/78267 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/72094 "Passed tests") | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/86/builds/16708 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/62379 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/78267 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/6658 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11112 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/47638 "Built successfully") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/48707 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/49996 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/48450 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

gsnedders commented on Tools/CISupport/Shared/de-duper:
```diff
@@ -0,0 +1,142 @@
+    args = parse_args(args)
+
+    args.root = os.path.abspath(args.root)
+    args.storage = os.path.abspath(args.storage) if args.storage[0] in ('.', '/') else os.path.join(args.root, args.storage)
```
This means that `./foo` and `foo` have different behaviour, which seems very surprising. What's the actual semantic you're trying to capture here?

JonWBedard replied:
`foo` would mean `/root/foo` where as `./foo` would mean `$cwd/foo`.

gsnedders replied:
That seems deeply surprising, and goes against how paths on the command-line work almost everywhere?

Like, I'd be okay with it always being relative to `args.root` like `os.path.join(args.root, args.storage)`, which would make `foo` and `./foo` both mean `/root/foo` (but `/storage/foo` would still work).

gsnedders commented on Tools/CISupport/Shared/de-duper:
```diff
@@ -0,0 +1,142 @@
+    processed = 0
+
+    existing_links = set()
+    for existing_file in os.listdir(args.storage):
```
We should be using `os.scandir`, then we don't need to have extra syscalls and potentially reopen the directory from disk to call `os.stat`.

gsnedders commented on Tools/CISupport/Shared/de-duper:
```diff
@@ -0,0 +1,142 @@
+
+    finally:
+        if process and process.poll() is None:
+            process.kill()
```
Why are we doing this rather than just using `os.walk` or similar? This opens us up to command-line injection for unclear benefit?

gsnedders commented on Tools/CISupport/Shared/de-duper:
```diff
@@ -0,0 +1,142 @@
+    )
+    if result.returncode:
+        return None
+    return result.stdout.split()[0]
```
Why are we doing this rather than using `hashlib.md5(usedforsecurity=False)`? (Also slightly surprised by the choice of md5; sha1 is typically quicker and what most other tools — not least git — use for content hashing.)

JonWBedard replied:
Validated that sha1 from Python is faster, using that.

gsnedders commented on Tools/CISupport/Shared/de-duper:
```diff
@@ -0,0 +1,142 @@
+        existing_file_abs_path = os.path.join(args.storage, existing_file)
+        if not is_linked(existing_file_abs_path):
+            os.remove(existing_file_abs_path)
+            cleaned_up += 1
```
I wonder if this should be behind a `--prune` flag or similar, rather than always doing this?

JonWBedard replied:
Removing these files isn't dangerous, though. If nothing is linked to them, they're dead. We'd just pass the `--prune` unconditionally.

gsnedders replied:
A broader dry-run option feels like it would be useful for testing this, at the very least? It _shouldn't_ be dangerous, that much is true, but it still feels non-zero risk!

gsnedders commented on Tools/CISupport/Shared/de-duper:
```diff
@@ -0,0 +1,142 @@
+    args.root = os.path.abspath(args.root)
+    args.storage = os.path.abspath(args.storage) if args.storage[0] in ('.', '/') else os.path.join(args.root, args.storage)
+    if not os.path.isdir(args.root):
+        sys.stderr.write(f"'{args.root}' is a path which does not exist\n")
```
```suggestion
        print(f"'{args.root}' is a path which does not exist", file=sys.stderr)
```

gsnedders commented on Tools/CISupport/Shared/de-duper:
```diff
@@ -0,0 +1,142 @@
+        return 1
+    storage_parent = os.path.dirname(args.storage)
+    if not os.path.isdir(storage_parent):
+        sys.stderr.write(f"'{storage_parent}' is a path which does not exist\n")
```
Also move to `print`?

gsnedders commented on Tools/CISupport/Shared/de-duper:
```diff
@@ -0,0 +1,142 @@
+    parser.add_argument(
+        'root', nargs='?',
+        help='Path to root of directory to de-duplicate PNG files in',
+        type=str, default='.',
```
It would be quite nice to use `pathlib.Path` as the type rather than raw strings everywhere.

bls1999 reviewed (approved):
[unofficial] rs=me

gsnedders commented on Tools/CISupport/Shared/de-duper:
```diff
@@ -0,0 +1,137 @@
+                    print(f'[dry-run] Storing and linking {full_path} to {sha_value}.png')
+                else:
+                    os.link(full_path, os.path.join(args.storage, f'{sha_value}.png'))
+                existing_links.add(sha_value)
```
This all seems wrong when `sha_value` is None; we probably just want an explicit `None` check and then continue? Probably logging something?

gsnedders commented on Tools/CISupport/Shared/de-duper:
```diff
@@ -0,0 +1,137 @@
+    args = parse_args(args)
+
+    args.root = os.path.abspath(args.root)
+    args.storage = os.path.abspath(args.storage) if args.storage[0] == '/' else os.path.join(args.root, args.storage)
```
There's no need to special-case `args.storage[0] == '/'`; `os.path.join` does the right thing itself.

JonWBedard committed:
[buildbot] Add de-duper script for results
https://bugs.webkit.org/show_bug.cgi?id=281334
rdar://137770187

Reviewed by Sam Sneddon.

Attempt to hard-link identical PNG files to a shared "storage" directory.

* Tools/CISupport/Shared/de-duper: Added.
(parse_args): Caller can specify root directory and storage directory.
(sha1): Compute the sha1 hash of the specified file.
(main): Iterate through all PNG files in the root directory, if those files
are already hard-links, skip them. Otherwise, compute the sha1 hash of the source
PNG files and move the file content to the storage directory and hard-link to that.

Canonical link: https://commits.webkit.org/285362@main
